Upscaling vignette (#93)
FRBCesab/funbiogeo: vignettes/upscaling.Rmd

---
title: "Upscaling your data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Upscaling your data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment  = "#>"
)
```

`funbiogeo` provides an easy way to upscale your site data to a coarser resolution.

> The idea is that you have any type of data at the site level (diversity metrics, 
> environmental data, etc.) but you would like to work on it at a coarser scale, 
> or you want to visualize it at that coarser scale. You may want to use the function
> `fb_aggregate_site_data()`.

Reviewer: ```suggestion
The idea is that you have any type of data at the site level (diversity metrics, 
environmental data, as well as site-species data) but you would like to work on it at a coarser scale, 
or you want to visualize it at that coarser scale. You may want to use the function
`fb_aggregate_site_data()`.
```

```{r setup}
library("funbiogeo")
```

> Let's import our site by locations object.

Reviewer: ```suggestion
Let's import our site by locations object, which describes the geographical locations of sites.
```

```{r import-site-locations}
data("site_locations")

site_locations
```

> These sites are a collection of regular spatial polygons at a resolution of 0.5°.

Reviewer: ```suggestion
These sites are a collection of regular spatial polygons at a resolution of 0.5° over Western Europe.
```

For each site, we want to compute the species richness. 

```{r compute-richness}
# Import site x species data ----
data("site_species")

# Compute species richness ----
species_richness <- fb_count_species_by_site(site_species)

head(species_richness)
```

Before going any further let's map these original values with the function
`fb_map_site_data()`.

```{r map-original-richness, dev='png'}
fb_map_site_data(site_locations, species_richness, "n_species")
```

Now, let's say that our next analyses require to work at a coarser resolution. We need to
aggregate site data on a new spatial grid (object `SpatRaster` from the 
[`terra`](http://cran.r-project.org/package=terra) package). Let's import this coarser raster.

```{r import-grid}
# Import study area grid ----
grd <- system.file("extdata", "grid_area.tif", package = "funbiogeo")
grd <- terra::rast(grd)

grd
```

> We will aggregate the site data (resolution of 0.5°) to this raster (resolution of 0.83°) 

Reviewer: ```suggestion
We will aggregate the site data (resolution of 0.5°) to this new coarser raster (resolution of 0.83°) 
```

with the function `fb_aggregate_site_data()`. This function requires the following arguments:

- `site_locations`: the site x locations object
- `site_data`: a `matrix` or `data.frame` containing values per sites to aggregate on the provided grid `agg_grid`. Can have one or several columns (variables to aggregate). The first column must contain sites names as provided in the example dataset `site_locations`
- `agg_grid`: a `SpatRaster` object (package `terra`). A raster of one single layer, that defines the grid along which to aggregate
- `fun`: the function used to aggregate sites values when there are multiple sites in one cell

Let's aggregate our species richness values on this grid.

```{r upscaled-richness}
# Upscale to grid ----
upscaled_richness <- fb_aggregate_site_data(
  site_locations = site_locations,
  site_data      = species_richness[ , 1:2],
  agg_grid       = grd,
  fun            = max
)

upscaled_richness
```

The result of this function is a `SpatRaster` where the cells contain the aggregated values of the species richness.

```{r map-upscaled-richness, dev='png'}
fb_map_raster(upscaled_richness)

> ```

Reviewer: I would add an example of aggregating site-species matrix here, as well as CWM (because we're in a functional perspective), but I can write this.

Author: Good idea. If you can add this example it could be great. Maybe w/ `fundiversity`/`mFD`?

Author: Thanks
